src/proofs/proof.rs

```diff
@@ -0,0 +1,514 @@
+use symbol_table::GlobalSymbol as Symbol;
+
+use crate::{
+    ast::{Command, Expr, Fact, GenericAction, GenericFact, Rule},
+    HashMap, HashSet, Term, TermDag,
+};
+
+type ProofId = usize;
+
+#[derive(Clone)]
+pub struct ProofStore {
+    store: Vec<ProofTerm>,
+    memo: HashMap<ProofTerm, ProofId>,
+    termdag: TermDag,
+}
+
+type Substitution = Vec<(Symbol, Term)>;
+
+fn subst_get(subst: &Substitution, sym: Symbol) -> Option<&Term> {
+    for ele in subst {
+        if ele.0 == sym {
+            return Some(&ele.1);
+        }
+    }
+    None
+}
+
+#[derive(Clone, PartialEq, Eq, Hash, Debug)]
+pub enum Proposition {
+    TOk(Term),
+    TEq(Term, Term),
+}
+
+// todo how to ignore this warning?
+#[warn(clippy::enum_variant_names)]
+#[derive(Clone, PartialEq, Eq, Hash)]
+pub enum ProofTerm {
+    /// proves a Proposition based on a rule application
+    /// the subsitution gives the mapping from variables to terms
+    /// the body_pfs gives proofs for each of the conditions in the query of the rule
+    /// the act_pf gives a location in the action of the proposition
+    PRule {
+        rule_name: Symbol,
+        subst: Substitution,
+        body_pfs: Vec<ProofId>,
+        result: Proposition,
+    },
+    /// A term is equal to itself- proves the proposition t = t
+    PRefl {
+        t_ok_pf: ProofId,
+        t: Term,
+    },
+    /// The symmetric equality of eq_pf
+    PSym {
+        eq_pf: ProofId,
+    },
+    PTrans {
+        pfxy: ProofId,
+        pfyz: ProofId,
+    },
+    /// get a proof for the child of a term given a proof of a term
+    PProj {
```

Comment:
Does this assume that rules only ever insert one term?

e.g. if I had a rule:

```
(rule ((foo x)) ((bar (Num x)) (baz x))
```
What does the proof of `(bar ...)` look like as opposed to `(baz ...)`. 

I take `PProj` to be helping with getting access to the `(Num ...)` subterm if the proof is of `bar`.

Reply:
I don't think we want to make that assumption... as far as I can tell, you can pick any term or sub-term in the rhs of a rule application when you use `PRule`.

That begs the question of why we need `PProj` at all. I'm not really sure, perhaps we won't need it?